### PR TITLE
[SPARK-40979][CORE] Keep removed executor info due to decommission

### PR DESCRIPTION
**Pull request by warrenzhu25**
### What changes were proposed in this pull request?
Keep removed executor info due to decommission in a separate map. Keep the max map size as 10000 to avoid excessive memory usage.

### Why are the changes needed?
Get info about whether executor is decommission even after this executor is removed

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Updated test in `TaskSchedulerImplSuite`

### CODE REVIEW

**Comment by warrenzhu25** [comment]
@dongjoon-hyun @Ngone51 Could you help take a look?

**Comment by mridulm** [comment]
We cannot keep this indefinitely - will cause OOM at driver.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Ya, +1 for @mridulm 's comment.

This topic should not be a follow-up PR because it needs a full review and irrelevant to the initial PR's goal.

I'm negative for this PR's approach for even an individual PR, @warrenzhu25 .

**Comment by dongjoon-hyun** [comment]
If the original contribution SPARK-40481 is insufficient for you still, we can revert it first before pursuing this kind of big change. We need to revisit from the bottom, @warrenzhu25 .

**Comment by warrenzhu25** [comment]
> If the original contribution [SPARK-40481](https://issues.apache.org/jira/browse/SPARK-40481) is insufficient for you still, we can revert it first before pursuing this kind of big change. We need to revisit from the bottom, @warrenzhu25 .

The original PR works for most cases, but some lost executors due to decommission cannot be ignored as we lost such info so I want to improve this. I understand the concern of keeping removed executor. 

What do you think of keeping only most recent 100 lost executors due to decommission in a separate set? @dongjoon-hyun  @mridulm

**Comment by dongjoon-hyun** [comment]
No, I don't think that is a correct way because there exists a large Spark cluster. Please don't try to remove this.

**Comment by warrenzhu25** [comment]
> No, I don't think that is a correct way because there exists a large Spark cluster. Please don't try to remove that line.

What I mean is something like below, the memory added is constant regardless of cluster size.

```
// limit the set size to 100 or 1000
val executorsRemovedDueToDecom = new HashSet[String]

executorsPendingDecommission.remove(executorId)
executorsRemovedDueToDecom.add(executorId)

```

**Comment by dongjoon-hyun** [comment]
Could you explain how it can prevent your problem completely? That sounds still greedy and best-effort approach to mitigate your issue.

> What I mean is something like below, the memory added is constant regardless of cluster size. @dongjoon-hyun
> 
> ```
> // limit the set size to 100 or 1000
> val executorsRemovedDueToDecom = new HashSet[String]
> 
> executorsPendingDecommission.remove(executorId)
> executorsRemovedDueToDecom.add(executorId)
> ```

**Comment by dongjoon-hyun** [comment]
As I mentioned before, this cannot be a followup. I created SPARK-40979 for you because you want to discuss this more.

**Comment by warrenzhu25** [comment]
> Could you explain how it can prevent your problem completely? That sounds still greedy and best-effort approach to mitigate your issue.
> 
> > What I mean is something like below, the memory added is constant regardless of cluster size. @dongjoon-hyun
> > ```
> > // limit the set size to 100 or 1000
> > val executorsRemovedDueToDecom = new HashSet[String]
> > 
> > executorsPendingDecommission.remove(executorId)
> > executorsRemovedDueToDecom.add(executorId)
> > ```

FetchFailed caused by decom executor could be divided into 2 categories:

1. When FetchFailed reached DAGScheduler, the executor is still alive or is lost but the lost info hasn't reached TaskSchedulerImpl. This is already handled in [SPARK-40979](https://issues.apache.org/jira/browse/SPARK-40979)
2. FetchFailed is caused by decom executor loss, so the decom info is already removed in TaskSchedulerImpl. If we keep such info in a short period, it is good enough. Even we limit the size of removed executors to 10K, it could be only at most 10MB memory usage. In real case, it's rare to have cluster size of over 10K and the chance that all these executors decomed and lost at the same time would be small.

**Comment by warrenzhu25** [comment]
@dongjoon-hyun @mridulm Updated pr to resolve your comments. Could you help take a look?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -2193,9 +2193,8 @@ private[spark] class DAGScheduler(
    * Return true when:
    *  1. Waiting for decommission start
    *  2. Under decommission process
-   * Return false when:
-   *  1. Stopped or terminated after finishing decommission
-   *  2. Under decommission process, then removed by driver with other reasons
+   *  3. Stopped or terminated after finishing decommission
+   *  4. Under decommission process, then removed by driver with other reasons
    */
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2198]: Reformulate the comment to reflect the recent changes ?
If decommissioned, will return `true` only until it expires.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 2]
```diff
@@ -145,6 +145,13 @@ private[spark] class TaskSchedulerImpl(
   // continue to run even after being asked to decommission, but they will eventually exit.
   val executorsPendingDecommission = new HashMap[String, ExecutorDecommissionState]
 
+  // Keep removed executors due to decommission, so getExecutorDecommissionState
+  // still return correct value even after executor is lost
+  val removedExecutorsDueToDecom = new mutable.LinkedHashMap[String, ExecutorDecommissionState]
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:150]: Instead of manually enforcing this manually, why not use CacheBuilder ?
Something like:
`CacheBuilder.newBuilder().maximumSize(10000).expireAfterAccess(1, TimeUnit.MINUTES).build()`

(Use suitable constants above)

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 3]
```diff
@@ -145,6 +145,13 @@ private[spark] class TaskSchedulerImpl(
...
+
+  // Max size to keep removed executors due to decommission
+  val REMOVED_EXECUTORS_MAX_SIZE = 10000
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:153]: Make this configurable, via an internal config - with default as 10k as here.
The total cluster size need not be large, but the number of executors can be higher (dynamic resource allocation, resource ask + size of node for rolling upgrades, spot instances coming in and out, etc).

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 4]
```diff
@@ -1117,7 +1124,14 @@ private[spark] class TaskSchedulerImpl(
...
 
-    executorsPendingDecommission.remove(executorId)
+    val removedDecomState = executorsPendingDecommission.remove(executorId)
+    removedDecomState.foreach(
+      s => {
+        removedExecutorsDueToDecom.put(executorId, s)
+        if (removedExecutorsDueToDecom.size >= REMOVED_EXECUTORS_MAX_SIZE) {
+          removedExecutorsDueToDecom.remove(removedExecutorsDueToDecom.head._1)
+        }
+      })
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:1134]:
```suggestion
    executorsPendingDecommission.remove(executorId).foreach(removedExecutorsDueToDecom.put(executorId, _))
```

Change to the Cache (above) will take care of eviction, etc.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 5]
```diff
@@ -2004,7 +2004,8 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext
 
     // executor1 is eventually lost
     scheduler.executorLost("executor1", ExecutorExited(0, false, "normal"))
-    assert(scheduler.executorsPendingDecommission.isEmpty)
+    // [SPARK-40481] after executor lost, decommission state still be kept
+    assert(scheduler.getExecutorDecommissionState("executor1").isDefined)
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:2008]: Also validate the cleanup behavior.
For example, set size to 2 and decom 3 executors to validate first got evicted - ensures size is bounded as code evovles.

**Review by mridulm — COMMENTED** [review]
I am fine with maintaining a cache of previously decomm'ed executors as long as they are cleaned up.

**Comment by mridulm** [comment]
Also, +CC @Ngone51

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 2 above)
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:150]: Also, rename variable to `executorsRemovedByDecom` or some such.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 6]
```diff
@@ -2024,6 +2024,15 @@ package object config {
       .stringConf
       .createOptional
 
+  private[spark] val SCHEDULER_MAX_RETAINED_REMOVED_EXECUTORS =
+    ConfigBuilder("spark.scheduler.maxRetainedRemovedExecutors")
+      .doc("Max number of removed executors by decommission to retain. This affects " +
+        "whether fetch failure caused by removed decommissioned executors could be ignored " +
+        "when spark.stage.ignoreDecommissionFetchFailure is enabled.")
+      .version("3.4.0")
+      .longConf
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2033]: `intConf`

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 7]
```diff
@@ -2024,6 +2024,15 @@ package object config {
       .stringConf
       .createOptional
 
+  private[spark] val SCHEDULER_MAX_RETAINED_REMOVED_EXECUTORS =
+    ConfigBuilder("spark.scheduler.maxRetainedRemovedExecutors")
+      .doc("Max number of removed executors by decommission to retain. This affects " +
+        "whether fetch failure caused by removed decommissioned executors could be ignored " +
+        "when spark.stage.ignoreDecommissionFetchFailure is enabled.")
+      .version("3.4.0")
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2032]: Mark this `internal` ?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 8]
```diff
@@ -145,6 +147,16 @@ private[spark] class TaskSchedulerImpl(
...
 
+  // Keep removed executors due to decommission, so getExecutorDecommissionState
+  // still return correct value even after executor is lost
+  val executorsRemovedByDecom =
+    CacheBuilder.newBuilder()
+      .maximumSize(conf.get(SCHEDULER_MAX_RETAINED_REMOVED_EXECUTORS))
+      .build[String, ExecutorDecommissionState]()
+
+  // Max size to keep removed executors due to decommission
+  val REMOVED_EXECUTORS_MAX_SIZE = 10000
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:158]: Remove this variable ?

**Review by mridulm — COMMENTED** [review]
Just a few minor comments, but mostly looks good to me.

+CC @dongjoon-hyun and @Ngone51 to also take a look.

**Comment by mridulm** [comment]
Can you address the remaining comment @warrenzhu25 ? Thx.
Once done, and tests pass, will wait for a few days to give Dongjoon a chance to review.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 9]
```diff
@@ -2024,6 +2024,16 @@ package object config {
       .stringConf
       .createOptional
 
+  private[spark] val SCHEDULER_MAX_RETAINED_REMOVED_EXECUTORS =
+    ConfigBuilder("spark.scheduler.maxRetainedRemovedExecutors")
+      .internal()
+      .doc("Max number of removed executors by decommission to retain. This affects " +
+        "whether fetch failure caused by removed decommissioned executors could be ignored " +
+        "when spark.stage.ignoreDecommissionFetchFailure is enabled.")
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2032]:
```suggestion
        s"when ${STAGE_IGNORE_DECOMMISSION_FETCH_FAILURE.key} is enabled.")
```

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 10]
```diff
@@ -2193,9 +2193,11 @@ private[spark] class DAGScheduler(
    * Return true when:
    *  1. Waiting for decommission start
    *  2. Under decommission process
-   * Return false when:
-   *  1. Stopped or terminated after finishing decommission
-   *  2. Under decommission process, then removed by driver with other reasons
+   *  3. Stopped or terminated after finishing decommission
+   *  4. Under decommission process, then removed by driver with other reasons
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2197]: This seems to change the original behavior. Would it affect any existing callers?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 11]
```diff
@@ -2024,6 +2024,16 @@ package object config {
       .stringConf
       .createOptional
 
+  private[spark] val SCHEDULER_MAX_RETAINED_REMOVED_EXECUTORS =
+    ConfigBuilder("spark.scheduler.maxRetainedRemovedExecutors")
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2028]:
```suggestion
    ConfigBuilder("spark.scheduler.maxRetainedRemovedDecommissionExecutors")
```

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 10 above)
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2197]: There is only one caller for this, right ?
We should make it `private` instead of `private[scheduler]` to remove the confusion.

Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2197]: Yes, upated.

**Review by mridulm — APPROVED** [review]
Looks good to me.
+CC @dongjoon-hyun , @Ngone51

**Review by mridulm — CHANGES_REQUESTED** [review]
Can you move `SCHEDULER_MAX_RETAINED_REMOVED_EXECUTORS` to below `STAGE_IGNORE_DECOMMISSION_FETCH_FAILURE` ?
This is causing the build failure.

**Comment by warrenzhu25** [comment]
> Can you move `SCHEDULER_MAX_RETAINED_REMOVED_EXECUTORS` to below `STAGE_IGNORE_DECOMMISSION_FETCH_FAILURE` ? This is causing the build failure.

Updated.

**Review by mridulm — APPROVED** [review]

**Review by Ngone51 — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 12]
```diff
@@ -2229,6 +2229,16 @@ package object config {
...
       .createWithDefault(false)
 
+  private[spark] val SCHEDULER_MAX_RETAINED_REMOVED_EXECUTORS =
+    ConfigBuilder("spark.scheduler.maxRetainedRemovedDecommissionExecutors")
+      .internal()
+      .doc("Max number of removed executors by decommission to retain. This affects " +
+        "whether fetch failure caused by removed decommissioned executors could be ignored " +
+        s"when ${STAGE_IGNORE_DECOMMISSION_FETCH_FAILURE.key} is enabled.")
+      .version("3.4.0")
+      .intConf
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2239]: Could you add the following?
```
.checkValue(_ >= 0, "needs to be a non-negative value")
```

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 13]
```diff
@@ -2229,6 +2229,16 @@ package object config {
...
+      .createWithDefault(10000)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2240]: Shall we start with `0` to avoid any regression?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 14]
```diff
@@ -2193,11 +2193,13 @@ private[spark] class DAGScheduler(
...
+   * Return false in case 3 and 4 when removed executors info are not retained.
+   * The max size of removed executors is controlled by
+   * spark.scheduler.maxRetainedRemovedExecutors
    */
-  private[scheduler] def isExecutorDecommissioningOrDecommissioned(
+  private def isExecutorDecommissioningOrDecommissioned(
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2202]: Shall we revert this change because this PR doesn't use `isExecutorDecommissioningOrDecommissioned`?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 15]
```diff
@@ -2004,17 +2006,30 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext
 
     // executor1 is eventually lost
     scheduler.executorLost("executor1", ExecutorExited(0, false, "normal"))
-    assert(scheduler.executorsPendingDecommission.isEmpty)
+    // [SPARK-40481] after executor lost, decommission state still be kept
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:2009]: Please remove `[SPARK-40481]`

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 16]
```diff
@@ -2004,17 +2006,30 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext
...
+    assert(scheduler.getExecutorDecommissionState("executor1").isDefined)
     // So now both the tasks are no longer running
-    assert(manager.copiesRunning.take(2) === Array(0, 0))
+
+    // executor 2 is decommissioned, then lost
+    scheduler.executorDecommission("executor2", ExecutorDecommissionInfo("", None))
+    scheduler.executorLost("executor2", ExecutorExited(0, false, "normal"))
+
+    // [SPARK-40481] only keep 1 removed executor info
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:2017]: Please remove `[SPARK-40481]`

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 17]
```diff
@@ -1976,10 +1977,11 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext
 
   test("test full decommissioning flow") {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:1978]: 1. After changing the default value `0`, please recover this test case.
2. Add a new test case with test prefix `SPARK-40979: ` with `SCHEDULER_MAX_RETAINED_REMOVED_EXECUTORS=1`.

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @warrenzhu25 . I left a few minor comments. The logic itself looks good.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 18]
```diff
@@ -2017,6 +2019,38 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext
     assert(manager.copiesRunning.take(2) === Array(1, 1))
   }
 
+  test("SPARK-40979: Test keep removed executor info due to decommissio") {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:2022]: Oh, typo.
`decommissio` -> `decommission`?

Comment (dongjoon-hyun) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:2022]:
```
- test("SPARK-40979: Test keep removed executor info due to decommissio") {
+ test("SPARK-40979: Keep removed executor info due to decommission") {
```

**Comment by mridulm** [comment]
Merged to master.
Thanks for working on this @warrenzhu25 !
Thanks for the reviews @dongjoon-hyun, @Ngone51 :-)
